components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAndOAuthAuthenticator.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.carbon.apimgt.gateway.handlers.security.authenticator;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.SynapseEnvironment;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityException;
+import org.wso2.carbon.apimgt.gateway.handlers.security.Authenticator;
+import org.wso2.carbon.apimgt.gateway.handlers.security.oauth.OAuthAuthenticator;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+/**
+ * Authenticator responsible for handling Mutual SSL and OAuth secured APIs.
+ */
+public class MutualSSLAndOAuthAuthenticator implements Authenticator {
```

Comment:
In the case of mutual ssl and oauth can't we use the OAuth and MutualSSL authenticators one after the other rather than having separate authenticator?

Reply:
We can use that. But when the user invokes the API every time, we need to go through an if else construct, to identify the api security for the relevant API. But if we have a separate authentictaor, the selection of authenticator is one time task only when publishing and we do not need to go through the check every time we invoke the API. WDYT?

Reply:
Even now, doesn't it go through this in each and every call? 

https://github.com/wso2/carbon-apimgt/blob/a379b19cf21a9b031b5767733bd95c8852308dcc/components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/APIAuthenticationHandler.java#L202

Reply:
Basically i am suggesting something similar below.
```
         if (isOAuthProtected) {
                authenticator = new OAuthAuthenticator(authorizationHeader, removeOAuthHeadersFromOutMessage);
authenticator.authenticate()
            } 
            if(isMutualSSLProtected) {
                authenticator = new MutualSSLAuthenticator(apiLevelPolicy, certificateInformation);
authenticator.authenticate()
            }
if(authenticator == null) {
// Give some error
}
```

OR

authenticator could be a list of authenticators

Reply:
@bhathiya  - It does not go through this particular segment every time, only when initializing the authentication handler, as I checked we go through the particular code segment only if authenticatior is null. Hence it will only be executed once in server startup time and when the relevant API is published again.

Reply:
@isharac - During the implementation phase, I also thought of both possibilities (i.e. Having list of authenticators in APIAuthenticationHandler and separate authenticator for combined authentication) and finally decided to go ahead with this design due to the reason I explained before and as well as considering following, we have recommended the customers to use the APIAuthenticationHandler as the extension point, if they want to go with the custom authentication method. In that case if we are introduce the list of authenticators,  For instance if the customer is extending and using this APIAuthenticationHandler, we cannot change the getAuthenticator method(with return type authenticator) as it is protected and the customer may be overriding it to initialize his authenticator and further we use that particular method when there is authentication failure. Hence to make sure not to break the backward compatibility I made sure to include a new authenticator. If this is an issue from maintainable perspective, for example if we are introduce a new type of API security mechanism, we may need to go with multiple authenticators (To include 2 secure mechanism, to include 3 secure mechanism). To avoid this shall I rename this as MultiAuthenticator, so that we will be able to use the same authenticator safely if we are to use different authentication mechanism in future.

Reply:
If the problem is back word compatibility, in the case of multiple authenticators what if we can have another method which will return list of enabled authenticators? And getAuthenticator() will return null.

So in the case of getAuthenticator()==null we can look for getAuthenticators()

At authentication failure we use this authenticator to get ChallengeString. In the case of  MutualSSLAndOAuthAuthenticator  you do a string concatenation to get this.
so we can do this in the APIAuthenticationHandler itself by looking at the list of authenticators.
wdyt?

Reply:
As discussed offline, we will be having a separate authenticator for multiple authentication support and will be changing the class in more generic way to support additional protocols.